### ISSUE TITLE
How to train GPRO on 2 GPUs, one for training, one for vllm

### ISSUE THREAD

**Issue by AIR-hl** [huggingface/trl#2864]
### Reproduction

When I use `Qwen2.5-3B-instruct` to train GRPO, the device for vllm always appear OOM when loading weights. II used two GPUs with 32GB of memory, one device for training, another for vllm. I dont know why a 3B model using so much memory on `device 1`

![Image](https://github.com/user-attachments/assets/79dfd03c-d123-496d-9fcc-07afc3027dff)

arguments settings:
```yaml
per_device_train_batch_size: 8
gradient_accumulation_steps: 8
num_generations: 8
use_vllm: true
vllm_gpu_memory_utilization: 0.8
use_peft: true
lora_r: 64
lora_alpha: 64
load_in_4bit: true
use_bnb_nested_quant: true
attn_implementation: flash_attention_2
bf16: true
...
```

Start command:
```shell
export CUDA_VISIBLE_DEVICES=0,1
accelerate launch --num_processes 1 train_Datawhale-R1.py --config Datawhale-R1.yaml
```

### System Info

- Platform: Linux-5.15.0-78-generic-x86_64-with-glibc2.35
- Python version: 3.10.8
- PyTorch version: 2.5.1
- CUDA device(s): NVIDIA vGPU-32GB, NVIDIA vGPU-32GB
- Transformers version: 4.48.3
- Accelerate version: 1.3.0
- Accelerate config: not found
- Datasets version: 3.1.0
- HF Hub version: 0.27.0
- TRL version: 0.16.0.dev0+ffcb9f4
- bitsandbytes version: 0.45.2
- DeepSpeed version: 0.16.3
- Diffusers version: 0.32.2
- Liger-Kernel version: not installed
- LLM-Blender version: not installed
- OpenAI version: 1.59.7
- PEFT version: 0.14.0

### Checklist

- [x] I have checked that my issue isn't already filed (see [open issues](https://github.com/huggingface/trl/issues?q=is%3Aissue))
- [x] I have included my system information
- [x] Any code provided is minimal, complete, and reproducible ([more on MREs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any code provided is properly formatted in code blocks, (no screenshot, [more on code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any traceback provided is complete

**Comment by MohamedAliRashad** [comment]
I have the same problem and i am using an 80GB GPU, how i am getting an OOM with such a small model ?

**Comment by zhzLuke96** [comment]
me too, OOM in `3b` model... Confusing

**Comment by leonardtang** [comment]
+1!

**Comment by Maghoumi** [comment]
See this [blog](https://www.philschmid.de/mini-deepseek-r1) to use a dedicated GPU for vLLM and use other GPUs for training

TLDR; run with `accelerate` with N-1 processes and don't forget to manually set `vllm_device='cuda:N'`.

**Comment by leonardtang** [comment]
I have 2 GPUs, and unfortunately running `accelerate launch --num_processes=0 train.py` and setting `vllm_device="cuda:1"` doesn't change the memory usage at all.

**Comment by Maghoumi** [comment]
Shouldn't that be `--num_processes=1`? Also, I did `pip install --upgrade git+https://github.com/huggingface/trl.git` on a fresh docker image based on `pytorch/pytorch:2.5.1-cuda12.4-cudnn9-devel` and it works for me.

**Comment by AIR-hl** [comment]
> Shouldn't that be `--num_processes=1`? Also, I did `pip install --upgrade git+https://github.com/huggingface/trl.git` on a fresh docker image based on `pytorch/pytorch:2.5.1-cuda12.4-cudnn9-devel` and it works for me.

Hi! I upgraded `trl` to the latest branch, but the `vllm` still appear OOM on device 1. I used a 3B model with lora and it can works well with unsloth on single GPU. When I reduce `vllm_gpu_memory_utilization` to 0.5, OOM was solved, but new error is here now

![Image](https://github.com/user-attachments/assets/208c60e8-a484-4190-8281-be0633aff668)

It seems that related to `bitandbbytes`, then I cancelled `qlora` and used `lora` instead, but OOM again. 
Anw, It's abnormal for a 3B model to experience OOM when inferring with `lora` and `vllm` on a GPU with 32GB of memory.

Part of my settings
```yaml
use_peft: true
lora_r: 64
lora_alpha: 64
# load_in_4bit: true
# use_bnb_nested_quant: true

use_vllm: true
vllm_device: cuda:1
vllm_gpu_memory_utilization: 0.8

torch_dtype: bfloat16
per_device_train_batch_size: 8
gradient_accumulation_steps: 8
num_generations: 8
```

```shell
export CUDA_VISIBLE_DEVICES=0,1
accelerate launch --num_processes 1 train_r1.py --config r1.yaml
```

**Comment by cieske** [comment]
I use same model as yours, and in my case I use one for vllm & two for training by `accelerate launch --multi-gpu --num-processes 2 ...` with just add `use_vllm=True` to GRPOConfig and no `CUDA_VISIBLE_DEVICES` options. With this settings training uses GPU 0, 1 and vllm uses GPU 2. 

OOM occured with rank0 GPU in your screenshot, is it really related to vllm? I experienced similar OOM(rank0, rakn1), using liger_kernel & gradient checkpointing, etc works.

**Comment by AIR-hl** [comment]
> I use same model as yours, and in my case I use one for vllm & two for training by `accelerate launch --multi-gpu --num-processes 2 ...` with just add `use_vllm=True` to GRPOConfig and no `CUDA_VISIBLE_DEVICES` options. With this settings training uses GPU 0, 1 and vllm uses GPU 2.
> 
> OOM occured with rank0 GPU in your screenshot, is it really related to vllm? I experienced similar OOM(rank0, rakn1), using liger_kernel & gradient checkpointing, etc works.

Hi! I think it is related to vllm or the correlated codes of trl, and there r few different errors,  this is just one of them.
It's weird that the device 1 wont OOM after set `vllm_gpu_memory_utilization=0.5`, but the utilization of device 1 is still more than `0.9` in reality. However, after OOM is solved, a new error happened that the tensor devices mismatch when gradient updating. 
Btw, 3GPU is completely different from 2 GPU settings, bcs mixed precision training is used on 3GPU settings(2 for training, 1 for vllm), but pure bf16 precision trainning on 2GPU settings (1 for training, 1 for vllm)
I already gave up, hufffft-_-

**Comment by zhzLuke96** [comment]
Hi, my OOM issue is resolved. Although I suspect it might be a bug with unsloth, it's very likely to be a bug related to trl, peft, or vllm on any of those sides.

My solution is, simply put, adding a small segment of `gc` code before inference. Everything seems to be working fine now, and I’ve trained for 3 epochs on my local machine without encountering any OOM issues, and the loss seems to be fine (appears normal).

You can check my comment in the unsloth issue here: https://github.com/unslothai/unsloth/issues/1744#issuecomment-2664908900

Alternatively, you could locate the specific place where `model.forward()` is called in your script and insert this before it:

```python
gc.collect()
torch.cuda.empty_cache()
```

I'm not sure if this will work for you (since you’re training based on trl), but it's worth a try. Memory leak issues can be quite difficult to handle and debug within the architecture, so this might be the simplest solution.

**Comment by cieske** [comment]
> tensor devices mismatch

I also meet devices mismatch error after overcome OOM, in my case load model w/ `AutoModelForCausalLM`, send it to `Accelerator().device` works.

Before
```
GRPOTrainer(model=model_path, ...)
```
After
```
accelerator = Accelerator()
device =  accelerator.device

model = AutoModelForCausalLM(model_path).to(device)

trainer = accelerator.prepare(GRPOTrainer(model=model, ...))
```

btw after solved mismatch error, I also suffer shape mismatch error....(weight must be 2-D, assert loaded_weight.shape[output_dim] == self.org_vocab_size)

**Comment by cieske** [comment]
> > I use same model as yours, and in my case I use one for vllm & two for training by `accelerate launch --multi-gpu --num-processes 2 ...` with just add `use_vllm=True` to GRPOConfig and no `CUDA_VISIBLE_DEVICES` options. With this settings training uses GPU 0, 1 and vllm uses GPU 2.
> > OOM occured with rank0 GPU in your screenshot, is it really related to vllm? I experienced similar OOM(rank0, rakn1), using liger_kernel & gradient checkpointing, etc works.
> 
> Hi! I think it is related to vllm or the correlated codes of trl, and there r few different errors, this is just one of them. It's weird that the device 1 wont OOM after set `vllm_gpu_memory_utilization=0.5`, but the utilization of device 1 is still more than `0.9` in reality. However, after OOM is solved, a new error happened that the tensor devices mismatch when gradient updating. Btw, 3GPU is completely different from 2 GPU settings, bcs mixed precision training is used on 3GPU settings(2 for training, 1 for vllm), but pure bf16 precision trainning on 2GPU settings (1 for training, 1 for vllm) I already gave up, hufffft-_-

I also meet devices mismatch error after overcome OOM, in my case load model w/ `AutoModelForCausalLM`, send it to `Accelerator().device` works.

Before
```
GRPOTrainer(model=model_path, ...)
```
After
```
accelerator = Accelerator()
device =  accelerator.device

model = AutoModelForCausalLM(model_path).to(device)

trainer = accelerator.prepare(GRPOTrainer(model=model, ...))
```

btw after solved mismatch error, I also suffer shape mismatch error....(weight must be 2-D, assert loaded_weight.shape[output_dim] == self.org_vocab_size)

**Comment by vagitablebirdcode** [comment]
The same isuue caused when use trl==0.15.2. I used two devices, one for training and one for vllm inference. when the first step is over, large vram is needed in  `logits = model(input_ids=input_ids, attention_mask=attention_mask, logits_to_keep=logits_to_keep + 1).logits` in `compute_loss` method and caused oom error.

**Comment by shing100** [comment]
+1

**Comment by Tuziking** [comment]
> > > I use same model as yours, and in my case I use one for vllm & two for training by `accelerate launch --multi-gpu --num-processes 2 ...` with just add `use_vllm=True` to GRPOConfig and no `CUDA_VISIBLE_DEVICES` options. With this settings training uses GPU 0, 1 and vllm uses GPU 2.我使用与您相同的模型，在我的情况下，我使用一个用于 vllm 和两个用于训练，通过 `accelerate launch --multi-gpu --num-processes 2 ...` 只需添加 `use_vllm=True` 到 GRPOConfig，没有 `CUDA_VISIBLE_DEVICES` 选项。使用这些设置，训练使用 GPU 0、1，而 vllm 使用 GPU 2。
> > > OOM occured with rank0 GPU in your screenshot, is it really related to vllm? I experienced similar OOM(rank0, rakn1), using liger_kernel & gradient checkpointing, etc works.内存溢出发生在您截图中的 rank0 GPU 上，这真的与 vllm 有关吗？我遇到了类似的 OOM（rank0，rakn1），使用 liger_kernel 和梯度检查点等方法有效。
> > 
> > 
> > Hi! I think it is related to vllm or the correlated codes of trl, and there r few different errors, this is just one of them. It's weird that the device 1 wont OOM after set `vllm_gpu_memory_utilization=0.5`, but the utilization of device 1 is still more than `0.9` in reality. However, after OOM is solved, a new error happened that the tensor devices mismatch when gradient updating. Btw, 3GPU is completely different from 2 GPU settings, bcs mixed precision training is used on 3GPU settings(2 for training, 1 for vllm), but pure bf16 precision trainning on 2GPU settings (1 for training, 1 for vllm) I already gave up, hufffft-_-嗨！我认为这与 vllm 或 trl 的相关代码有关，并且有几个不同的错误，这只是其中之一。很奇怪，设备 1 在设置 `vllm_gpu_memory_utilization=0.5` 后不会发生 OOM，但实际上设备 1 的利用率仍然高于 `0.9` 。然而，在解决 OOM 问题后，梯度更新时出现了新的错误，即张量设备不匹配。顺便说一下，3GPU 的设置与 2GPU 的设置完全不同，因为在 3GPU 设置中使用了混合精度训练（2 个用于训练，1 个用于 vllm），而在 2GPU 设置中（1 个用于训练，1 个用于 vllm）纯 bf16 精度训练我已经放弃了，唉-_-
> 
> I also meet devices mismatch error after overcome OOM, in my case load model w/ `AutoModelForCausalLM`, send it to `Accelerator().device` works.我也在解决 OOM 问题后遇到了设备不匹配错误，在我的情况下，使用 `AutoModelForCausalLM` 加载模型，发送到 `Accelerator().device` 可以工作。
> 
> Before  在
> 
> ```
> GRPOTrainer(model=model_path, ...)
> ```
> 
> After  在
> 
> ```
> accelerator = Accelerator()
> device =  accelerator.device
> 
> model = AutoModelForCausalLM(model_path).to(device)
> 
> trainer = accelerator.prepare(GRPOTrainer(model=model, ...))
> ```
> 
> btw after solved mismatch error, I also suffer shape mismatch error....(weight must be 2-D, assert loaded_weight.shape[output_dim] == self.org_vocab_size)btw 解决了不匹配错误后，我也遇到了形状不匹配错误....（权重必须是二维的，断言 loaded_weight.shape[output_dim] == self.org_vocab_size）

I also encountered this problem, did you solve it?

**Comment by vagitablebirdcode** [comment]
I find that the following function in GRPOTrainer cause the vram peak sudden uppper:
```python
def _get_per_token_logps(self, model, input_ids, attention_mask, logits_to_keep):
        # We add 1 to `logits_to_keep` because the last logits of the sequence is later excluded
        logits = model(input_ids=input_ids, attention_mask=attention_mask, logits_to_keep=logits_to_keep + 1).logits
        logits = logits[:, :-1, :]  # (B, L-1, V), exclude the last logit: it corresponds to the next token pred

        input_ids = input_ids[:, -logits_to_keep:]
        # For transformers<=4.48, logits_to_keep argument isn't supported, so here we drop logits ourselves.
        # See https://github.com/huggingface/trl/issues/2770
        logits = logits[:, -logits_to_keep:]
        return selective_log_softmax(logits, input_ids)  #  compute logprobs for the input tokens
```
I try to split the logits to many small chunk as follows:
```python
def _get_per_token_logps(self, model, input_ids, attention_mask, logits_to_keep):
        chunk_size = 1
        batch_size = input_ids.shape[0]
        all_logps = []

        for i in range(0, batch_size, chunk_size):
            chunk_input_ids = input_ids[i:i+chunk_size]
            chunk_attention_mask = attention_mask[i:i+chunk_size]

            logits = model(
                input_ids=chunk_input_ids,
                attention_mask=chunk_attention_mask,
                logits_to_keep=logits_to_keep + 1
            ).logits

            logits = logits[:, :-1, :]
            chunk_input_ids = chunk_input_ids[:, -logits_to_keep:]
            logits = logits[:, -logits_to_keep:]

            chunk_logps = selective_log_softmax(logits, chunk_input_ids)
            all_logps.append(chunk_logps)

            del logits
            torch.cuda.empty_cache()

        return torch.cat(all_logps, dim=0)
```
then it works. Every one can change the `chunk size` according to the GPU vram.

**Comment by albertvillanova** [comment]
Hey @AIR-hl, sorry for the very late reply — we're catching up on the backlog. 

This issue predates a full redesign of TRL's vLLM integration, so the exact setup you hit no longer applies: the `vllm_device` argument and the "launch with `--num_processes N-1` and pin vLLM to `cuda:N`" pattern have both been removed. vLLM now runs in one of two documented modes — colocate (shares the training GPUs, the default) or server (a separate `trl vllm-serve` process on its own GPUs); see https://github.com/huggingface/trl/blob/7c1fa85ad5d1878edef7a795c74e91749961b152/docs/source/vllm_integration.md#L277-L279 and the "best distributed setup" guidance in the same doc. On the peak-memory spike a couple of you traced to `_get_per_token_logps`, the reference/old log-prob passes are now chunked internally (https://github.com/huggingface/trl/blob/7c1fa85ad5d1878edef7a795c74e91749961b152/trl/trainer/grpo_trainer.py#L1434), and there are several new knobs for peak memory (chunked cross-entropy, `ds3_gather_for_generation=False`, vLLM sleep mode) in https://github.com/huggingface/trl/blob/7c1fa85ad5d1878edef7a795c74e91749961b152/docs/source/reducing_memory_usage.md. 

I'm closing this as outdated, but please open a fresh issue with a current `trl env` if you still OOM on a recent version — happy to dig in.
